lib/logstash-docket/repository.rb

```diff
@@ -0,0 +1,177 @@
+# encoding: utf-8
+
+require_relative 'util/threadsafe_index'
+require_relative 'util/threadsafe_deferral'
+
+require_relative 'artifact_plugin'
+require_relative 'source'
+require_relative 'rubygem_info'
+
+module LogstashDocket
+  ##
+  # A `Repository` contains the complete source and release history of a named artefact
+  # that is pushed to rubygems. It has a {@link Source}, which allows us to extract contents
+  # of the versioned artefacts, and provides methods for listing releases.
+  #
+  class Repository
```

Comment:
I will merge as-is and follow up with a refactor, since we need this to be stable for the 7.5 release and the automated refactor touched way too much for me to be comfortable.